### ISSUE TITLE
Neo4j slow startup time for large database (1200GB, 60mins+)

### ISSUE THREAD

**Issue by in3rsha** [neo4j/neo4j#12150]
I have a large Neo4j database that is roughly 1200GB. The average startup time when running `sudo neo4j start` is around 60 minutes (or more).

Michael Hunger asked if I could share some details about my startup times. Here's my server and Neo4j config:

```
OS  = Ubuntu 16.04
CPU = CPU: Intel Xeon X5675 (24) @ 3.068GHz
RAM = 128GB (32GB swap)
SSD = 4000GB
```

```
dbms.memory.heap.initial_size=16g
dbms.memory.heap.max_size=16g
dbms.memory.pagecache.size=64g
```

And here's the output from `/var/log/neo4j/neo4j.log` for my most recent starts of Neo4j:

**1. 110 mins** (after a hard-kill)

```
2019-01-02 13:22:39.792+0000 INFO  Starting...
2019-01-02 13:22:41.431+0000 INFO  Initiating metrics...
2019-01-02 15:06:24.884+0000 INFO  Deprecated index providers in use: lucene-1.0 (6 indexes). Use procedure 'db.indexes()' to see what indexes use which index provider.
2019-01-02 15:06:25.283+0000 INFO  Sending metrics to CSV file at /var/lib/neo4j/metrics
2019-01-02 15:06:25.787+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2019-01-02 15:06:27.774+0000 INFO  Started.
2019-01-02 15:06:28.491+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.all)
2019-01-02 15:06:28.492+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.idle)
2019-01-02 15:06:28.616+0000 INFO  Mounted REST API at: /db/manage
2019-01-02 15:06:28.696+0000 INFO  Server thread metrics have been registered successfully
2019-01-02 15:06:29.718+0000 INFO  Remote interface available at http://localhost:7474/
```

**2. 62 mins**

```
2019-01-14 13:28:51.773+0000 INFO  ======== Neo4j 3.5.0 ========
2019-01-14 13:28:51.779+0000 INFO  Starting...
2019-01-14 13:28:53.187+0000 INFO  Initiating metrics...
2019-01-14 14:30:46.354+0000 INFO  Deprecated index providers in use: lucene-1.0 (6 indexes). Use procedure 'db.indexes()' to see what indexes use which index provider.
2019-01-14 14:30:46.780+0000 INFO  Sending metrics to CSV file at /var/lib/neo4j/metrics
2019-01-14 14:30:47.211+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2019-01-14 14:30:49.237+0000 INFO  Started.
2019-01-14 14:30:49.802+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.all)
2019-01-14 14:30:49.802+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.idle)
2019-01-14 14:30:49.917+0000 INFO  Mounted REST API at: /db/manage
2019-01-14 14:30:49.998+0000 INFO  Server thread metrics have been registered successfully
2019-01-14 14:30:51.005+0000 INFO  Remote interface available at http://localhost:7474/
```

**3. 53 mins** (After Upgrading to 3.5.1) (also changed from NewRatio=2 to NewRatio=8)

```
2019-01-17 14:13:02.410+0000 INFO  ======== Neo4j 3.5.1 ========
2019-01-17 14:13:02.414+0000 INFO  Starting...
2019-01-17 14:13:03.680+0000 INFO  Initiating metrics...
2019-01-17 15:06:04.788+0000 INFO  Deprecated index providers in use: lucene-1.0 (6 indexes). Use procedure 'db.indexes()' to see what indexes use which index provider.
2019-01-17 15:06:05.136+0000 INFO  Sending metrics to CSV file at /var/lib/neo4j/metrics
2019-01-17 15:06:05.567+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2019-01-17 15:06:07.415+0000 INFO  Started.
2019-01-17 15:06:08.021+0000 INFO  Mounted REST API at: /db/manage
2019-01-17 15:06:08.108+0000 INFO  Server thread metrics have been registered successfully
2019-01-17 15:06:09.093+0000 INFO  Remote interface available at http://localhost:7474/
```

**4. 71 mins** 

```
2019-01-22 10:38:34.979+0000 INFO  ======== Neo4j 3.5.1 ========
2019-01-22 10:38:34.984+0000 INFO  Starting...
2019-01-22 10:38:36.392+0000 INFO  Initiating metrics...
2019-01-22 11:49:47.628+0000 INFO  Deprecated index providers in use: lucene-1.0 (6 indexes). Use procedure 'db.indexes()' to see what indexes use which index provider.
2019-01-22 11:49:48.108+0000 INFO  Sending metrics to CSV file at /var/lib/neo4j/metrics
2019-01-22 11:49:48.577+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2019-01-22 11:49:50.556+0000 INFO  Started.
2019-01-22 11:49:51.132+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.all)
2019-01-22 11:49:51.133+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.idle)
2019-01-22 11:49:51.165+0000 INFO  Mounted REST API at: /db/manage
2019-01-22 11:49:51.284+0000 INFO  Server thread metrics have been registered successfully
2019-01-22 11:49:52.565+0000 INFO  Remote interface available at http://localhost:7474/
```

I've also noticed that when Neo4j is starting, the server doesn't seem to be under much stress in terms of CPU usage. None of the cores are running at (or close to) 100%.

I was wondering if there is anything I could do to speed up the start time.

**Comment by chrisvest** [comment]
Try profiling with FlightRecorder/Java Mission Control, or VisualVM, to see what it's doing. It'd be useful if you could share those profiles. Also try setting the log level to debug: https://neo4j.com/docs/operations-manual/current/reference/configuration-settings/#config_dbms.logs.debug.level

**Comment by in3rsha** [comment]
Thanks @chrisvest. I haven't profiled using FlightRecorder, but I set `dbms.logs.debug.level=DEBUG` in `neo4j.conf` and restarted Neo4j.

It looks like the delay is taking place during `Initiating metrics...`, but unfortunately doesn't give any detailed information about what's going on:

**5. 44 mins**

```
2019-01-30 19:48:57.876+0000 INFO  ======== Neo4j 3.5.2 ========
2019-01-30 19:48:57.886+0000 INFO  Starting...
2019-01-30 19:48:59.208+0000 INFO  Initiating metrics...
2019-01-30 20:32:11.741+0000 INFO  Deprecated index providers in use: lucene-1.0 (6 indexes). Use procedure 'db.indexes()' to see what indexes use which index provider.
2019-01-30 20:32:16.796+0000 INFO  Sending metrics to CSV file at /var/lib/neo4j/metrics
2019-01-30 20:32:18.020+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2019-01-30 20:32:20.000+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.all)
2019-01-30 20:32:20.021+0000 WARN  Server thread metrics not available (missing neo4j.server.threads.jetty.idle)
2019-01-30 20:32:22.665+0000 INFO  Started.
2019-01-30 20:32:23.273+0000 DEBUG Adding JAXRS classes [org.neo4j.server.rest.MasterInfoService, org.neo4j.server.rest.CoreDatabaseAvailabilityService, org.neo4j.server.rest.ReadReplicaDatabaseAvailabilityService, org.neo4j.server.rest.causalclustering.CausalClusteringService] at [/db/manage]
2019-01-30 20:32:23.273+0000 INFO  Mounted REST API at: /db/manage
2019-01-30 20:32:23.280+0000 DEBUG Adding JAXRS classes [org.neo4j.server.rest.discovery.DiscoveryService] at [/]
2019-01-30 20:32:23.280+0000 DEBUG Adding JAXRS classes [org.neo4j.server.rest.dbms.UserService] at [/]
2019-01-30 20:32:23.289+0000 DEBUG Adding JAXRS classes [org.neo4j.server.rest.web.RestfulGraphDatabase, org.neo4j.server.rest.web.TransactionalService, org.neo4j.server.rest.web.CypherService, org.neo4j.server.rest.web.DatabaseMetadataService, org.neo4j.server.rest.web.ExtensionService, org.neo4j.server.rest.web.ResourcesService, org.neo4j.server.rest.web.BatchOperationService] at [/db/data]
2019-01-30 20:32:23.293+0000 DEBUG Adding JAXRS classes [org.neo4j.server.rest.management.JmxService, org.neo4j.server.rest.management.RootService, org.neo4j.server.rest.management.VersionAndEditionService] at [/db/manage]
2019-01-30 20:32:23.294+0000 DEBUG Adding JAXRS classes [org.neo4j.server.rest.management.console.ConsoleService] at [/db/manage]
2019-01-30 20:32:23.438+0000 INFO  Server thread metrics have been registered successfully
2019-01-30 20:32:23.467+0000 DEBUG Mounting servlet at [/db/manage]
2019-01-30 20:32:23.505+0000 DEBUG Mounting servlet at [/db/data]
2019-01-30 20:32:23.563+0000 DEBUG Mounting servlet at [/]
2019-01-30 20:32:24.981+0000 INFO  Remote interface available at http://localhost:7474/
``` 

Looks like I'll need to profile using FlightRecorder to get more information.

A quick question whilst I've got you... I have no experience with the Java ecosystem and only have command-line access to the server. Is FlightRecorder easy enough to install for a Java novice (i.e. achievable in a day)?

**Comment by chrisvest** [comment]
If you use Oracle JDK 8, then you can instruct the JVM to grab a flight recording after a certain delay from the startup, and have it dump the result to a path of your choice: https://docs.oracle.com/javacomponents/jmc-5-4/jfr-runtime-guide/run.htm#JFRUH164

Flight Recorder is built into the Oracle JDK. It is not included in Open JDK (technically, it is from Java 11 onwards, but don't use that just yet).

Note that you are only allowed to do this in your testing or development environments, unless you have a Java license from Oracle. That's what the `-XX:+UnlockCommercialFeatures` means.

Use the `dbms.jvm.additional` settings in `neo4j.conf` to add the parameters to the Neo4j JVM: https://neo4j.com/docs/operations-manual/current/reference/configuration-settings/#config_dbms.jvm.additional

Also add the following two JVM options, to improve the accuracy of the profiling: `-XX:+UnlockDiagnosticVMOptions` `-XX:+DebugNonSafepoints`.

**Comment by in3rsha** [comment]
### **EDIT: Fixed it.** Just changed the path to `/tmp/myrecording.jfr`

Thanks. Just getting a small error and thought I'd ask here as I couldn't find anything after a quick search.

I've added these settings to `/etc/neo4j/neo4j.conf`:

```
dbms.jvm.additional=-XX:+UnlockCommercialFeatures
dbms.jvm.additional=-XX:+FlightRecorder
dbms.jvm.additional=-XX:StartFlightRecording=duration=60s,filename=myrecording.jfr
dbms.jvm.additional=-XX:+UnlockDiagnosticVMOptions
dbms.jvm.additional=-XX:+DebugNonSafepoints
```

I'm seeing when I run `sudo service neo4j start`:

```
Feb 05 11:25:53 me neo4j[11781]: Starting Neo4j.
Feb 05 11:25:54 me neo4j[11781]: Could not start recording, not able to write to file "myrecording.jfr"
```

Sorry to have to ask, but do you have any suggestions on fixing this **not able to write to file** error?

**Comment by chrisvest** [comment]
The command you have will start a recording as soon as the JVM is able, I think. Since your startup is taking upwards 40 minutes, it'd also be nice to have a recording from later into the startup, like 20 minutes or so. You can use `jcmd $PID JFR.start duration=60s filename=/tmp/myrecording-2.jfr` to do that.

**Comment by in3rsha** [comment]
Thanks again. I just did a stop/start under normal conditions and Neo4j loaded in **1 minute**. Typical!

It seems as though the slow startup happens when I have to hard-kill Neo4j because the server is suffering from memory problems (i.e. swap memory fills up and the server starts thrashing). This is a separate issue of mine where too much of the Neo4j `pagecache` gets moved to swap memory and exceeds the swap memory available.

Anyway, I've left the FlightRecorder in place and will post the results the next time I have restart Neo4j due to memory problems.

**Comment by jexp** [comment]
Please configure your heap plus pagecache so that it doesn't exceed the physical memory.

**Comment by chrisvest** [comment]
If you have time, you can try giving it a `kill -9` to simulate a generic crash, and see if it's the recovery that generally takes a long time.

**Comment by in3rsha** [comment]
> Please configure your heap plus pagecache so that it doesn't exceed the physical memory.

Hi Michael (@jexp). I have `128GB` of RAM, and these are my settings:

```
dbms.memory.heap.initial_size=16g
dbms.memory.heap.max_size=16g
dbms.memory.pagecache.size=64g
```

Over time, I find that some of the Neo4j `pagecache` gradually gets moved to swap memory until it fills up completely (32GB). Ideally I'd like to stop this from happening, but I don't know how. Here is the thread I posted: https://community.neo4j.com/t/is-it-possible-to-reduce-the-amount-of-swap-space-that-neo4j-uses/4735

Do you have any ideas on what might be causing this?

**Comment by in3rsha** [comment]
### Minor Update

As is typical with testing, I haven't been able to recreate the slow startup time just yet. Here are my failed attempts so far:

**1. Neo4j `stop` and `start` when the server is running smoothly.**

Neo4j starts up in around **1 minute**.

**2. `kill -9` when the server is running smoothly.**

Killing Neo4j when the server isn't having any memory problems doesn't seem to affect the startup time.

Neo4j starts up in around **1 minute**.

**3. Force server to run out of RAM.**

I configured Neo4j to use more pagecache than is available in RAM, and also ran a script that would eat up memory so that the server would eventually run out. Neo4j happily shut itself down when the memory ran out, and after I freed up memory, started up again in around **1 minute**.

**4. [ ] Next: Allow Neo4j to use too much swap memory.**

It seems I run in to the slow startup time when I have configured Neo4j appropriately, but over the course of time the `pagecache` gets moved to swap memory and causes the server to start thrashing. So Neo4j has enough memory available to run, but the server is swapping so much that it slows down to a crawl.

It's at this point that I find it almost impossible to shutdown Neo4j, and have to `kill -9` it or restart the server. After that, I get the 60 minute+ startup time.

Anyway, I have configured the server so that it will run in to this scenario again at some point, but it might take a few days for the swap to fill up. I will post back in a few days with the results!

**Comment by jexp** [comment]
Thanks a lot Greg for all your efforts

**Comment by jexp** [comment]
Thanks a lot Greg for all your efforts

**Comment by sherfert** [comment]
Hi @in3rsha 
Do you feel it would be OK to close this issue until you are able to reproduce the slow startup and provide a .jfr file for us to analyze?

**Comment by in3rsha** [comment]
> Hi @in3rsha
> Do you feel it would be OK to close this issue until you are able to reproduce the slow startup and provide a .jfr file for us to analyze?

Yes. I haven't been able to reproduce this slow start since I opened this issue (typical!).

My best guess would be that whatever was causing the slow start up after killing Neo4j due to memory issues was fixed in the upgrade to Neo4j 3.5.2+.

I have left the FlighRecorded in place just in case it happens again. Thank you to everyone for all the help.
